src/Microsoft.OData.Core/UriParser/Resolver/UnqualifiedODataUriResolver.cs

```diff
@@ -47,27 +47,39 @@ public override IEnumerable<IEdmOperation> ResolveBoundOperations(IEdmModel mode
                 return base.ResolveBoundOperations(model, identifier, bindingType);
             }
 
-            return FindAcrossModels<IEdmOperation>(model, identifier, this.EnableCaseInsensitive)
-                .Where(operation =>
-                    operation.IsBound
-                    && operation.Parameters.Any()
-                    && operation.HasEquivalentBindingType(bindingType));
+            return FindOperations(model, identifier, this.EnableCaseInsensitive, true, bindingType);
         }
 
-        private static IEnumerable<T> FindAcrossModels<T>(IEdmModel model, String qualifiedName, bool caseInsensitive) where T : IEdmSchemaElement
+        private static IEnumerable<IEdmOperation> FindOperations(IEdmModel model, string qualifiedName, bool caseInsensitive, bool isBound = false, IEdmType bindingType = null)
         {
-            Func<IEdmModel, IEnumerable<T>> finder = (refModel) =>
-                refModel.SchemaElements.OfType<T>()
-                .Where(e => string.Equals(qualifiedName, e.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal));
+            IList<IEdmOperation> results = new List<IEdmOperation>();
+
+            StringComparison strComparison = caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
 
-            IEnumerable<T> results = finder(model);
+            GetOperationsForModel(model, qualifiedName, results, isBound, bindingType, strComparison);
 
             foreach (IEdmModel reference in model.ReferencedModels)
             {
-                results.Concat(finder(reference));
+                GetOperationsForModel(reference, qualifiedName, results, isBound, bindingType, strComparison);
             }
 
             return results;
         }
+
+        private static void GetOperationsForModel(IEdmModel model, string qualifiedName, IList<IEdmOperation> results, bool isBound, IEdmType bindingType, StringComparison strComparison)

```

Comment:
>GetOperationsForModel [](http://example.com/codeflow?start=28&length=21)

IEdmModel has a FindDeclaredBoundOperations that is already cached in the base model implementation.  Rather than loop through all schema elements (which, in a service like MS Graph, is huge), why not just call FindDeclaredBoundOperations and iterate through that list, yielding operations with the right name?

Reply:
Ohh I wonder if that expensive O(n) search (loop through all schema elements) will explain most of the insane cost we are seeing from this callstack in AGS ...

Reply:
yes; findboundoperations should be hugely faster if we don't have to loop through all schema elements.

It's unfortunate that we don't have a FindDeclaredUnboundOperations in IEdmModel. We should add one for V8.x. In the meantime, we could consider adding something to the EdmModelBase and attempt to cast the IEdmModel to EdmModelBase for the common case.